text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1020 @@
+to users, since Ember has not had any modifiers besides `{{action}}`, and the
+element modifier manager RFC was just merged recently. Key things to focus on
+here will be the fact the modifiers run whenever a node is inserted or removed,
+so they may run multiple times during a components life (if they exist in an
+`{{if}}` or `{{each}}` block, for instance), and that they can receive pure
+function helpers if they do not require any state beyond the state of the
+element itself, which can simplify component code.
+
+### Converting From Classic Components
+
+For users who are familiar with the classic component API, some changes will be
+fairly simple, and others will be more nuanced. The individual focus points
+should be those outlined in the motivation of this RFC, and the motivation for
+the changes should be included in a summary that helps users understand the
+intent of new APIs so they can use them effectively.
+
+**Outer HTML semantics** should be fairly simple to communicate. These should
+also be the easiest to codemod, since it is mostly a 1-to-1 conversion from the
+classic component definition to the template.
+
+**Namespaced arguments** will be a sticking point for some. The main thing to
+focus on here is aliasing to provide defaults, and the benefits of immutability.
+Patterns like the `@arg` decorator in
+[sparkles-decorators](https://github.com/gossi/sparkles-decorators) make this
+fairly ergonomic, and for simpler components defaulting using a class field
+initializer can also work.
+
+**Lifecycle hooks** may be fairly confusing to existing users, who are used to
+having a variety of hooks to choose from when coordinating lifecycle events. The
+bullet points here are:
+
+* Tracked properties/computed properties are the primary place to react to
+  argument changes for any values that can be computed directly via getters.
+  Ideally, most logic for derived state is conventionally in tracked or computed
+  properties.
+* `didUpdateArgs` is the place to react to argument changes that don't fit into
+  getters. This includes asynchronous calls, conditional changes (e.g. if an arg
+  changed, do X, else do nothing), etc.
+* `willDestroy` is the correct place for all teardown code, like in classic
+  components.
+* `didReceiveAttrs`, `willRender`, and `didRender` code should be extracted into
+  functions which are then passed to `{{did-render}}`
+* `willDestroyElement` code should be extracted into functions which are then
+  passed to `{{will-destroy}}`
+
+**Element modifiers** will be new to almost everyone, and this will have to be
+a primary focus in learning. The bullet points for these are:
+
+* They are conceptually like the render lifecycle hooks on classic components,
+  but they run on elements instead of components.
+* `{{did-render}}` is similar to the `didRender` lifecycle hooks in classic
+  components. It will run when the element is inserted, and then it will run
+  again when any of its _args_ change.
```

Comment:
In moving rendering lifecycle hooks out of the JS/TS module and into the template (attached to elements), we make development of element-less components more challenging.

Example: `ember-google-maps` allows a declarative google maps configuration to be built in handlebars w/ syntax like

```hbs
{{#g-map lat='51.508530' lng='-0.076132' zoom=10 as |g|}}
  {{#each locations as |l|}}
    {{g.marker lat=l.lat lng=l.lng onClick=(action 'showDetails' l)}}
  {{/each}}
{{/g-map}}
```
Only the outer-most component renders anything (a single element -- a `<div>` I think) -- all of its children serve the purpose a declarative DSL for describing JavaScript calls to the gmaps API.

How will projects like these implement post-render behavior for these inner-components while having no DOM elements?

Reply:
Looking at `ember-google-maps` code, there is exactly one use of `didRender`, which is for the overlay components. This does seem like a valid use case, but based on this and the code in `ember-leaflet` I don't believe it is that very common as is, and it is not the case we should be optimizing for, even for element-less components.

In cases where users _do_ need this functionality, such as the overlay, I think it would be perfectly reasonable to use a custom component manager to add a `didRender` hook. This could even be distributed as a `render-detector` component, like the one being used in `ember-google-maps`.

Reply:
I actually think this would be a better suited job for a [ResizeObserver](https://developers.google.com/web/updates/2016/10/resizeobserver) actually, the main thing they want is to detect when the content's size has changed, not when it has rerendered.